src/backend/cdb/cdbmutate.c

```diff
@@ -2666,44 +2670,45 @@ root_slice_is_executed_on_coordinator(Plan *plan, PlannerInfo *root)
 			 * distribution policy is not master-only, then the root slice is
 			 * not executed on the coordinator.
 			 */
-			return false;
+			result = false;
 		}
 	}
 
 	if (plan->flow)
 	{
 		Flow	   *flow = plan->flow;
 
+		if (flow->flotype != FLOW_SINGLETON || flow->segindex >= 0)
+		{
+			/*
+			 * For non-singleton or singleton on segment, the root slice is
+			 * not executed on the coordinator.
+			 */
+			result = false;
+		}
+
 		if (root->glob->is_parallel_cursor)
 		{
-			if (flow->flotype != FLOW_SINGLETON ||
-				(flow->locustype != CdbLocusType_Entry &&
-				 flow->locustype != CdbLocusType_General &&
-				 flow->locustype != CdbLocusType_SingleQE))
+			if (flow->flotype == FLOW_SINGLETON &&
+				(flow->locustype == CdbLocusType_Entry ||
+				 flow->locustype == CdbLocusType_General ||
+				 flow->locustype == CdbLocusType_SingleQE))
 			{
 				/*
-				 * For these scenarios, parallel retrieve cursor doesn't need
-				 * to run on coordinator, since endpoint QE doesn't need to
+				 * For these scenarios, parallel retrieve cursor needs
+				 * to run on coordinator, since endpoint QE needs to
 				 * interact with the retrieve connections.
 				 */
-				return false;
+				result = true;
```

Comment:
So here we can overwrite `result` by true if it was set to false somewhere upper. Is it correct?
What was the reason of this commit? Why it's important to use variable instead of return?

Reply:
> So here we can overwrite `result` by true if it was set to false somewhere upper. Is it correct?

Yes, as I understand it, the executor does just that. https://github.com/arenadata/gpdb/blob/ea3c4baf4c7b25d4020e1b5e5bd83d95e9c894c1/src/backend/executor/execMain.c#L1809-L1843

Reply:
> Yes, as I understand it, the executor does just that.
It's a good answer with the link. I don't need to break my mind searching for a reason.
Still better to comment this in time and not on demand, but anyway, thank you. I'll take a look.

Reply:
The funny thing is that I have not yet found a test to confirm the need for all this.

Reply:
And in general, I believe that the executor, not the planner, should check the plan for validity, because if the planner understands that the plan is invalid, then why shouldn't it rebuild the plan?